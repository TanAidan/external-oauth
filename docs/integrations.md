Integrations in Pizzly are derived from individual JSON configuration files, located in the [`/integrations`](https://github.com/Bearer/Pizzly/tree/master/integrations) directory at the root of the Pizzly project. For example, the GitHub integration is configured at [`/integrations/github.json`](https://github.com/Bearer/Pizzly/blob/master/integrations/github.json) and looks like this:

```json
{
  "name": "GitHub",
  "auth": {
    "authorizationURL": "https://github.com/login/oauth/authorize",
    "tokenURL": "https://github.com/login/oauth/access_token",
    "authType": "OAUTH2",
    "tokenParams": {},
    "authorizationParams": {},
    "auth": { "accessType": "offline" }
  },
  "request": {
    "baseURL": "https://api.github.com/",
    "headers": {
      "Accept": "application/vnd.github.v3+json",
      "Authorization": "token ${auth.accessToken}",
      "User-Agent": "Pizzly"
    }
  }
}
```

All requests sent to the third-party APIs are made based on the content of the relevant configuration file.

## Supported API integrations

There are dozens of APIs pre-configured with Pizzly ([full list here](/docs/supported-apis.md). But if the API you are looking for is not listed yet, you can create a new JSON configuration file with the name of the API.

## Adding a new integration

Adding a new integration to Pizzly involves two parts: Creating the integration JSON config file, and importing the new config into the integrations manifest at `/integrations/index.ts`.

### The integration JSON file

Each API has a configuration JSON file associated with it. The file is located within the `/integrations/` directory and is named afterward the API name (in lowercase).

| API Name | Configuration file name                                                                | OAuth Type |
| -------- | -------------------------------------------------------------------------------------- | ---------- |
| GitHub   | [github.json](https://github.com/Bearer/Pizzly/blob/master/integrations/github.json)   | OAuth 2.0  |
| Reddit   | [reddit.json](https://github.com/Bearer/Pizzly/blob/master/integrations/reddit.json)   | OAuth 2.0  |
| Twitter  | [twitter.json](https://github.com/Bearer/Pizzly/blob/master/integrations/twitter.json) | OAuth 1.0a |

In case the API name has several words, use an hyphen between each word (e.g [google-calendar.json](https://github.com/Bearer/Pizzly/blob/master/integrations/google-calendar.json)).

### The integration structure

Each JSON file has four top-level keys: `name`, `image`, `auth`, and `request`.

```jsonc
{
  // The name of the integration
  "name": "API Name",

  // Image URL for API logo (optional)
  "image": "https://path.to.image.url.jpg"

  // All authorization-related settings
  "auth": {}

  // Defines how to perform API requests
  "request": {}
}
```

#### `name` and `image`

The name and image keys define how the API will be displayed in the Pizzly dashboard. Image is optional, and will be picked up by Clearbit's image API if omitted. If you notice an image not showing up adding an image URL is recommended. (As can be seen in the Google Calendar integration file)

#### `auth` (for Authentication)

The next main section is the `auth` key. This object contains all the necessary information needed for Pizzly to perform the OAuth dance with the API provider. These values can often be found at the authentication page of the API provider's documentation. Some are specific to the authentication type.

- `authType`: The authentication type, either `OAUTH1` or `OAUTH2`.
- `authorizationURL` (OAUTH2): The URL that users are redirected to when asked to authorize your application. Usually the provider's auth endpoint, such as `https://github.com/login/oauth/authorize`.
- `tokenURL` (OAUTH2): The URL used to exchange the auth code for the token during the auth flow. For example: `https://github.com/login/oauth/access_token`.
- `requestTokenURL` (OAUTH1): The URL provided by the API for retrieving the request token.
- `accessTokenURL` (OAUTH1): The URL provided by the API for retrieving the access token.
- `userAuthorizationURL` (OAUTH1): The URL provided by the API for authorizing the user.
- `signatureMethod` (OAUTH1): The method used by the API to sign requests.
- `authorizationParams`: This object contains any static parameters that need to be passed in during authorization. For example, the Discord API expects `"response_type": "code"` and `"grant_type": "authorization_code"` to be sent during authorization. Pizzly will send common dynamic values back to the API.
- `tokenParams`: Similar to the `authorizationParams` key, this object contains any required static parameters that need to be sent to the token URL. For example, GitLab expects the `response_type` with the authorization URL, but `grant_type` with the token URL.

In cases where `authorizationParams` or `tokenParams` are not required, they can be left as empty objects or omitted completely.

```jsonc
{
  "name": "API Name",
  "image": "https://path.to.image.url.jpg",

  // All authorization-related settings
  "auth": {
    // This is the URL that user's will be directed to when
    // initiating the authorization flow.
    "authorizationURL": "https://example.api/oauth",

    // The URL used for exchanging the auth code for a token
    "tokenURL": "https://example.api/oauth/token",

    // Set to either OAUTH1 or OAUTH2
    "authType": "OAUTH2",

    // Any fixed parameters that the API expects during token exchange
    "tokenParams": {
      // The grant_type is a common token parameter
      "grant_type": "authorization_code"
    },

    // Set any required, fixed, parameters to be passed during authorization
    // ex. state, access_type, prompt, etc
    "authorizationParams": {
      "response_type": "code"
    }
  }
}
```

#### `request`

The final part of the config file is the `request` key. It takes the `baseURL`, an object of `headers`, and an object of `params`. This is also where you will need to take advantage of the variable interpolations that Pizzly offers. You may have seem them in the other config files, such as `${auth.accessToken}`. These properties are dependant on the API provider and auth type. For example:

```jsonc
{
  "name": "API Name",
  "image": "https://path.to.image.url.jpg",
  "auth": {},
  "request": {
    // The API base URL
    "baseURL": "https://api.example.com/",

    // Extra headers that will be passed with each request
    "headers": {
      "Accept": "application/json",
      "Authorization": "Bearer ${auth.accessToken}",
      "User-Agent": "Pizzly"
    },

    // Extra parameters that will be passed in the body of each API call
    "params": {
      "example": "value"
    }
  }
}
```

## Interpolations

Some APIs require extra information than Pizzly supports per se. To support this APIs, the configuration file can use extra variable interpolations.

You might have notice the `${auth.accessToken}` which is used in all integration's configurations. Others available interpolation objects are the `${headers.[...]}` and `${connectParams.[...]}`.

This can be used for APIs that require more information than Pizzly currently supports. For example, an API that requires a custom base URL can be accessed by passing a header with each API call. Pizzly's proxy will make sure to replace the variable from the configuration with the value provided in the header, e.g. : `"baseURL":"https://${headers.api_account_name}.example.com"`

## Add the integration to the `index.ts` file in `/integrations`

The `index.ts` file in the integrations folder imports all integrations when the Pizzly server starts. To begin testing and using your new integration, add an import statement to `index.ts`.

## Testing your configuration

Make sure that your integration works locally. To start testing a new integration, follow the instructions for [Running Pizzly on your machine](/docs/readme.md). Confirm that Pizzly is working locally by visiting `http://localhost:8080` before attempting to add a new integration.

## Further configuration

Most settings are described in this page, but we recommend using an existing API's configuration file that matches the auth type of the new API.
