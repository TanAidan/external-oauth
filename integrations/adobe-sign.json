{
  "name": "Adobe Sign",
  "image": "https://logo.clearbit.com/adobe.com",
  "auth": {
    "authorizationURL": "https://secure.eu1.echosign.com/public/oauth",
    "tokenURL": "https://api.eu1.echosign.com/oauth/token",
    "refreshURL": "https://api.eu1.echosign.com/oauth/refresh",
    "authType": "OAUTH2"
  },
  "request": {
    "baseURL": "https://api.eu1.echosign.com/api/rest/v6/",
    "headers": {
      "Accept": "application/json",
      "Authorization": "Bearer ${auth.accessToken}",
      "User-Agent": "Pizzly"
    }
  }
}
