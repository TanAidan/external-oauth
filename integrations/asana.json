{
  "name": "Asana",
  "auth": {
    "authorizationURL": "https://app.asana.com/-/oauth_authorize",
    "tokenURL": "https://app.asana.com/-/oauth_token",
    "authType": "OAUTH2",
    "tokenParams": { "grant_type": "authorization_code" },
    "authorizationParams": {},
    "auth": { "response_type": "code" }
  },
  "request": {
    "baseURL": "https://app.asana.com/api/1.0/",
    "headers": {
      "Accept": "application/json",
      "Authorization": "Bearer ${auth.accessToken}",
      "User-Agent": "Pizzly"
    }
  }
}
