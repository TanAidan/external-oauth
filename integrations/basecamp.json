{
  "name": "Basecamp",
  "auth": {
    "authorizationURL": "https://launchpad.37signals.com/authorization/new",
    "tokenURL": "https://launchpad.37signals.com/authorization/token",
    "authType": "OAUTH2",
    "tokenParams": {
      "grant_type": "authorization_code",
      "type": "web_server"
    },
    "authorizationParams": { "type": "web_server" },
    "auth": { "response_type": "code" }
  },
  "request": {
    "baseURL": "https://3.basecampapi.com/${headers.basecamp_account_id}/",
    "headers": {
      "Accept": "application/json",
      "Authorization": "Bearer ${auth.accessToken}",
      "User-Agent": "Pizzly"
    }
  }
}
