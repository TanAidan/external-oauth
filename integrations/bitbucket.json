{
  "name": "BitBucket",
  "auth": {
    "authorizationURL": "https://bitbucket.org/site/oauth2/authorize",
    "tokenURL": "https://bitbucket.org/site/oauth2/access_token",
    "authType": "OAUTH2",
    "tokenParams": { "grant_type": "authorization_code" },
    "authorizationParams": {},
    "auth": { "response_type": "code" }
  },
  "request": {
    "baseURL": "https://api.bitbucket.org/2.0/",
    "headers": {
      "Accept": "application/json",
      "Authorization": "Bearer ${auth.accessToken}",
      "User-Agent": "Pizzly"
    }
  }
}
