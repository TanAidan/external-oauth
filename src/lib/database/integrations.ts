/**
 * Basic operations with the ./integrations folder
 * to get or list the integrations within that folder.
 */

import fs from 'fs'
import path from 'path'
import { Types } from '../../types'

const integrationsDir = path.join(__dirname, '../../../', 'integrations')

/**
 * List all integrations within the folder
 * @returns an array of the integrations configuration.
 */

export const list = async (): Promise<Types.Integration[]> => {
  return new Promise((resolve, reject) => {
    fs.readdir(integrationsDir, function(err, integrationsFiles) {
      if (err) {
        return reject('Unable to access integrations directory.')
      }

      const apis: any[] = []
      integrationsFiles.forEach(function(file) {
        const fileExtension = file.split('.').pop()
        if (fileExtension !== 'json') {
          return
        }

        try {
          const fileName = file.slice(0, -5)
          const fileContent = require(path.join(integrationsDir, `${fileName}.json`))
          const integration = formatIntegration(fileName, fileContent)

          apis.push(integration)
        } catch (err) {}
      })

      resolve(apis)
    })
  })
}

/**
 * Retrieve a particular integration within the folder.
 * @param apiName
 * @returns the integration configuration.
 */

export const get = async (integrationName: string): Promise<Types.Integration> => {
  return new Promise((resolve, reject) => {
    if (!integrationName) {
      return reject('Empty integration name provided.')
    }

    try {
      const fileContent = require(path.join(integrationsDir, `${integrationName}.json`))
      const integration = formatIntegration(integrationName, fileContent)
      return resolve(integration)
    } catch (err) {
      return reject(err)
    }
  })
}

const formatIntegration = (fileName: string, fileContent: any) => {
  const integration = fileContent as Types.Integration
  integration.id = fileName
  integration.image =
    integration.image || 'https://logo.clearbit.com/' + integration.name.toLowerCase().replace(' ', '') + '.com'

  const isOAuth2Auth = isOAuth2(integration)
  integration.auth.setupKeyLabel = isOAuth2Auth ? 'Client ID' : 'Consumer Key'
  integration.auth.setupSecretLabel = isOAuth2Auth ? 'Client Secret' : 'Consumer Secret'

  return integration
}

/**
 * Validation
 */

export const validateConfigurationScopes = (scopesAsString: string): string[] | null => {
  const scopes: string = ((String(scopesAsString) as string) || '').trim()

  return (scopes && scopes.split(/\r?\n/)) || null
}

export const validateConfigurationCredentials = (
  setup: { [key: string]: string } | undefined,
  integration: Types.Integration
): Types.OAuth1Credentials | Types.OAuth2Credentials | undefined => {
  if (!setup) {
    return
  }

  const authConfig = integration.auth
  const isOAuth2 = authConfig.authType == 'OAUTH2'
  const isOAuth1 = authConfig.authType == 'OAUTH1'

  if (isOAuth1) {
    const consumerKey = String(setup.consumerKey)
    const consumerSecret = String(setup.consumerSecret)

    if (consumerKey && consumerSecret) {
      return { consumerKey, consumerSecret }
    }
  } else if (isOAuth2) {
    const clientId = String(setup.clientId)
    const clientSecret = String(setup.clientSecret)

    if (clientId && clientSecret) {
      return { clientId, clientSecret }
    }
  }

  return
}

/*
  Helpers
*/

export function isOAuth2(integration: Types.Integration): integration is Types.Integration<Types.OAuth2Config> {
  return integration.auth.authType === 'OAUTH2'
}

export function isOAuth1(integration: Types.Integration): integration is Types.Integration<Types.OAuth1Config> {
  return integration.auth.authType === 'OAUTH1'
}
