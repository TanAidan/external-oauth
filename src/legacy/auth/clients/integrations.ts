// import { SetupDetailsNotFound } from '../errors'
import { TIntegrationConfig, EAuthType } from '../v3/types'
import '../../../../integrations'
import Knex from 'knex'

interface ICommonOAuthPayload {
  serviceName: string
  userId: string
  updatedAt: number
  setupId: string
  scopes?: any
  tokenResponseJSON?: string
  callbackParamsJSON?: string
  connectParams?: any
}

export type TOAuth1Payload = ICommonOAuthPayload & {
  accessToken: string
  tokenSecret: string
  consumerKey?: string
  consumerSecret?: string
  expiresIn: number
}

export type TOAuth2Payload = ICommonOAuthPayload & {
  accessToken: string
  clientId?: string
  clientSecret?: string
  refreshToken?: string
  expiresIn?: number
  idToken?: string
  idTokenJwt?: any
}

interface IAuthParams {
  buid: string
  authId: string
}

export type TOAuthPayload = TOAuth1Payload | TOAuth2Payload

export interface IAuthResult {
  id: string
  buid: string
  auth_id: string
  payload: {
    accessToken: string
    refreshToken: string
    idToken: string
    expiresIn: any
    scopes: string[]
    tokenResponseJSON: string
    updatedAt: any
    idTokenJwt: any
    connectParams?: any
    callbackParamsJSON: any
  }
}

export const updateAuth = async ({
  buid,
  authId,
  setupId,
  payload,
  store
}: IAuthParams & { payload: TOAuthPayload } & { store: any; setupId: string }) => {
  await store('authentications').insert({ buid, auth_id: authId, setup_id: setupId, payload })
}

export const getAuth = async <IAuthResult>({ buid, authId, store }: IAuthParams & { store: any }) => {
  console.log('[getAuth] buid/authId', buid, authId)

  return (await store('authentications')
    .where({ buid, auth_id: authId })
    .first()) as IAuthResult
}

export const getConfig = async ({ buid }: { buid: string }) => {
  let item = {} as any
  try {
    item = require(`../../../../integrations/${buid}.json`)
  } catch (err) {
    if (err.code === 'MODULE_NOT_FOUND') {
      return false
    }
  }

  const configItem = item.auth as TIntegrationConfig
  if (configItem.authType) {
    configItem.authType = configItem.authType.toUpperCase() as EAuthType
  }

  return { ...configItem, authConfig: configItem, requestConfig: item.request }
}

export const saveSetupDetails = async ({
  buid,
  setupId,
  credentials,
  scopes,
  store
}: {
  buid: string
  setupId: string
  store: Knex
  credentials: any
  scopes: string[] | undefined
}) => {
  await store('configurations').insert({ buid, credentials, scopes, setup_id: setupId })
}

export const getSetupDetails = async ({
  buid,
  setupId,
  store
}: {
  buid: string
  setupId: string | undefined
  store: Knex
}) => {
  if (setupId) {
    return await store('configurations')
      .where({ buid, setup_id: setupId })
      .first()
  }

  return await store('configurations')
    .where({ buid })
    .orderBy('updated_at', 'desc')
    .first()
}
