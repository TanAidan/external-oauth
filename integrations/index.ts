import './adobe-sign.json'
import './asana.json'
import './basecamp.json'
import './bitbucket.json'
import './coinbase.json'
import './cronofy.json'
import './digital-ocean.json'
import './discord.json'
import './docusign.json'
import './drift.json'
import './dropbox.json'
import './eventbrite.json'
import './facebook.json'
import './fitbit.json'
import './freshbooks.json'
import './freeagent.json'
import './front.json'
import './full-contact.json'
import './github.json'
import './gitlab.json'
import './goodreads.json'
import './google-calendar.json'
import './google-contacts.json'
import './google-docs.json'
import './google-drive.json'
import './google-drive-upload.json'
import './google-mail.json'
import './google-sheets.json'
import './google-analytics.json'
import './google-hangouts.json'
import './help-scout.json'
import './hootsuite.json'
import './hubspot.json'
import './imgur.json'
import './intercom.json'
import './intuit.json'
import './instagram.json'
import './jira.json'
import './linkedin.json'
import './mailchimp.json'
import './microsoft-one-drive.json'
import './microsoft-one-note.json'
import './microsoft-teams.json'
import './moltin.json'
import './monday.json'
import './online.json'
import './oura.json'
import './pagerduty.json'
import './pinterest.json'
import './pipedrive.json'
import './platform_sh.json'
import './product-hunt.json'
import './quickbooks-sandbox.json'
import './quickbooks.json'
import './reddit.json'
import './rescuetime.json'
import './revolut-sandbox.json'
import './sage.json'
import './salesforce.json'
import './sellsy.json'
import './shopify.json'
import './shutterstock.json'
import './slack.json'
import './smart-sheet.json'
import './snov.json'
import './soundcloud.json'
import './spotify.json'
import './square.json'
import './strava.json'
import './survey-monkey.json'
import './todoist.json'
import './trello.json'
import './twist.json'
import './twitch.json'
import './twitter.json'
import './typeform.json'
import './uber.json'
import './unsplash.json'
import './vimeo.json'
import './whatsapp.json'
import './xero.json'
import './youtube.json'
import './zendesk-chat.json'
import './zoho-books.json'
import './zoho-crm.json'
import './zoho-invoices.json'
import './zoho-subscriptions.json'
import './zoom.json'
import './splitwise.json'
import './youneedabudget.json'
