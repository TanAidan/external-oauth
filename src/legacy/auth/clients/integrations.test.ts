// @ts-ignore
import { getConfig, getSetupDetails, updateAuthV3, revokeAuthV3, getAuth, saveSetupDetails } from './integrations'

describe('integration client', () => {
  // const servicesTableName = 'test-services-table'
  // const setupId = 'a-setup-id'
  // const clientId = 'test-client-id'
  // const authId = 'test-auth-id'
  const buid = 'test-buid'

  describe('getConfig', () => {
    const config = { test: 'config' }

    it('retrieves the config from DynamoDb', async () => {
      const result = await getConfig({ buid })

      expect(result).toBe(config)
    })

    it("raises an InvalidBuid error if the config doesn't exist", async () => {
      expect(getConfig({ buid })).rejects.toMatchSnapshot()
    })

    it('uppercases the authType', async () => {
      // const result = await getConfig({ buid })
      // expect(result.config.authType).toBe('OAUTH1')
    })
  })

  // describe('getSetupDetails', () => {
  //   const details = { test: 'details' }
  //   const data = { data: 'value' }
  //   // const detailsWithData = { data, ...details }
  //   const scopedUserDataTableName = 'test-scoped-data-table'

  //   it('retrieves the config from DynamoDb', async () => {
  //     const result = await getSetupDetails({ clientId, buid, setupId, scopedUserDataTableName })

  //     expect(result).toBe(details)
  //   })

  //   it('returns the `data` attribute when one exists, rather than the whole item', async () => {
  //     const result = await getSetupDetails({ clientId, buid, setupId, scopedUserDataTableName })

  //     expect(result).toBe(data)
  //   })

  //   it("raises a SetupDetailsNotFound error if the config doesn't exist", async () => {
  //     expect(getSetupDetails({ clientId, buid, setupId, scopedUserDataTableName })).rejects.toMatchSnapshot()
  //   })
  // })
})
