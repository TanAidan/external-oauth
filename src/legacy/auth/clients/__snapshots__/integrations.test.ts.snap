// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration client getConfig retrieves the config from DynamoDb 1`] = `[MockFunction]`;
